Add Minio hosting/client integration

Harold-Morgan opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- ⚠️ We will not merge the PR to the main repo if your changes are not in a *feature branch* of your forked repository. Create a new branch in your repo, **do not use the `main` branch in your repo**! -->

<!-- ⚠️ **Every PR** needs to have a linked issue and have previously been approved. PRs that don't follow this will be rejected. >

<!-- 👉 It is imperative to resolve ONE ISSUE PER PR and avoid making multiple changes unless the changes interrelate with each other -->

<!-- 📝 Please always keep the "☑️ Allow edits by maintainers" button checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork. This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->

**Closes #606 **

<!-- Add an overview of the changes here -->

I've added maybe somewhat barebone (but working!) MiniO storage and client (based on the official https://github.com/minio/minio-dotnet library). It was a ton of fun.

<!-- All details should be in the linked issue. Feel free to call out any outstanding differences here. -->

I've added 

```
SkipIfKeyedRegistrationIsNotSupported(useKey);
```

to ConformanceTests.cs/ConnectionInformationIsDelayValidated because this integration doesn't support keyed registrations untill https://github.com/minio/minio-dotnet/issues/1147 is resolved. I hope this isn't too much, but maybe it should've been done in separate PR.

## PR Checklist

<!-- Please check if your PR fulfills the following requirements, and remove the ones that are not applicable to the current PR -->

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [x] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [x] New integration
  - [x] Docs are written
  - [x] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
- [x] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Every new API (including internal ones) has full XML docs
- [x] Code follows all style conventions

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->

## Other information

<!-- Please add any other information that might be helpful to reviewers. -->

Harold-Morgan commented:
@dotnet-policy-service agree

ctolkien commented:
Would love to see this get finished off.

davidfowl commented on src/CommunityToolkit.Aspire.Hosting.Minio/MinioBuilderExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+        string name,
+        IResourceBuilder<ParameterResource>? rootUser = null,
+        IResourceBuilder<ParameterResource>? rootPassword = null,
+        int minioPort = 9000,
```
```suggestion
        int? minioPort = null,
```

Expose a WithHostPort, WithPassword and WithUsername

Harold-Morgan replied:
Thanks!

I was looking at the integrations in this repo, but this got me looking at a Postgres integration at the main Aspire repo, and I've tweaked this and few things to the better

davidfowl commented on src/CommunityToolkit.Aspire.Hosting.Minio/MinioBuilderExtensions.cs:
```diff
@@ -0,0 +1,76 @@
+        IResourceBuilder<ParameterResource>? rootUser = null,
+        IResourceBuilder<ParameterResource>? rootPassword = null,
+        int minioPort = 9000,
+        int minioConsolePort = 9001)
```
Remove, this is usually fixed

davidfowl commented:
This looks really good! Left a few comments.

davidfowl commented:
> may not have been such a good idea (here's a [link](https://github.com/CommunityToolkit/Aspire/pull/691/files#diff-6c096f8bdb8d9a0a52199e8d769e7e5b863a29236681e836a190a4574cabffc3R95) for further reference)

It's a good idea. Just expose all of the parts of the connection string so the users can make their own connection strings, but having a default is good.

davidfowl commented on tests/CommunityToolkit.Aspire.Testing/TestDistributedApplicationBuilder.cs:
```diff
@@ -79,7 +79,7 @@ void Configure(DistributedApplicationOptions applicationOptions, HostApplication
             cfg.AddInMemoryCollection(new Dictionary<string, string?>
             {
-                ["DcpPublisher:RandomizePorts"] = "true",
+                ["DcpPublisher:RandomizePorts"] = "false",
```
For all tests?

Harold-Morgan replied:
I can't comprehend how it got here, honestly. Maybe I forgot it while debugging ports. What a shame :(

Fxd.

Harold-Morgan commented:
Now this looks somewhat ready to review, lifting draft

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Minio/CommunityToolkit.Aspire.Hosting.Minio.csproj:
```diff
@@ -0,0 +1,15 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <AdditionalPackageTags>MinioS3 Hosting integration</AdditionalPackageTags>
```
```suggestion
    <AdditionalPackageTags>minio hosting cloud storage</AdditionalPackageTags>
```

Harold-Morgan replied:
Fixed, thanks

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Minio/CommunityToolkit.Aspire.Hosting.Minio.csproj:
```diff
@@ -0,0 +1,15 @@
+
+  <PropertyGroup>
+    <AdditionalPackageTags>MinioS3 Hosting integration</AdditionalPackageTags>
+    <Description>A .NET Aspire hosting integration for MiniO</Description>
```
```suggestion
    <Description>A .NET Aspire hosting integration for MinIO</Description>
```

Harold-Morgan replied:
Fixed, thanks

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Minio/MinioBuilderExtensions.cs:
```diff
@@ -0,0 +1,173 @@
+    }
+    
+    /// <summary>
+    /// Configures the host port that the PGAdmin resource is exposed on instead of using randomly assigned port.
```
PGAdmin?

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Minio/MinioContainerResource.cs:
```diff
@@ -0,0 +1,81 @@
+    /// <param name="name">The name of the resource.</param>
+    /// <param name="user">A parameter that contains the Minio server root user name.</param>
+    /// <param name="password">A parameter that contains the Minio server root password.</param>
+    public MinioContainerResource(string name, ParameterResource user, ParameterResource password) : base(name)
+    {
+        RootUser = user;
+        PasswordParameter = password;
+    }
```
Can you use a primary constructor for this

Harold-Morgan replied:
Yes, it looks much cleaner now, thanks!

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Minio/MinioContainerResource.cs:
```diff
@@ -0,0 +1,81 @@
+        return builder.Build();
+    }
+
+    internal void SetPassword(ParameterResource password)
+    {
+        PasswordParameter = password;
+    }
```
Since there is a public `set` on the `PasswordParameter`, does this method make sense? Should instead the property have a non-public `set`?

Harold-Morgan replied:
I do believe that this method looks cleaner in builder than explicitly setting the property

But I do agree that public set doesn't make much sense combined with this method. So I followed your advice and set it to private, thanks!

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Minio/README.md:
```diff
@@ -0,0 +1,37 @@
+# CommunityToolkit.Aspire.Hosting.MinoO library
```
```suggestion
# CommunityToolkit.Aspire.Hosting.MinIO library
```

aaronpowell commented on src/CommunityToolkit.Aspire.Minio.Client/CommunityToolkit.Aspire.Minio.Client.csproj:
```diff
@@ -0,0 +1,17 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <AdditionalPackageTags>MinioS3 Client integration</AdditionalPackageTags>
```
```suggestion
    <AdditionalPackageTags>minio client cloud storage</AdditionalPackageTags>
```

aaronpowell commented on src/CommunityToolkit.Aspire.Minio.Client/CommunityToolkit.Aspire.Minio.Client.csproj:
```diff
@@ -0,0 +1,17 @@
+
+  <PropertyGroup>
+    <AdditionalPackageTags>MinioS3 Client integration</AdditionalPackageTags>
+    <Description>A .NET Aspire client integration for MinioS#</Description>
```
```suggestion
    <Description>A .NET Aspire client integration for MinioS3</Description>
```

aaronpowell commented on src/CommunityToolkit.Aspire.Minio.Client/MinioClientBuilderExtensionMethods.cs:
```diff
@@ -0,0 +1,100 @@
+    private const string DefaultConfigSectionName = "Aspire:Minio:Client";
+
+    /// <summary>
+    /// Adds Minio Client to ASPNet host
```
```suggestion
    /// Registers <see cref="IMinioClient" /> as a singleton in the services provided by the <paramref name="builder"/>.
```

Or something to that effect.

aaronpowell commented on src/CommunityToolkit.Aspire.Minio.Client/MinioClientBuilderExtensionMethods.cs:
```diff
@@ -0,0 +1,100 @@
+                client.SetTraceOff();
+        }
+
+        var minioClientFactory = new MinioClientFactory(ConfigureClient);
+        builder.Services.TryAddSingleton<IMinioClientFactory>(minioClientFactory);
+        
+        IMinioClient GetClient()
+        {
+            if (settings.Endpoint is null)
+                throw new InvalidOperationException("The MiniO endpoint must be provided either in configuration section, or as a part of connection string or settings delegate");
+            
+            return minioClientFactory.CreateClient();
+        }
+        
+        switch (settings.ServiceLifetime)
+        {
+            case ServiceLifetime.Singleton:
+                builder.Services.TryAddSingleton(_ => GetClient());
+                break;
+            case ServiceLifetime.Scoped:
+                builder.Services.TryAddScoped(_ => GetClient());
+                break;
+            case ServiceLifetime.Transient:
+                builder.Services.TryAddTransient(_ => GetClient());
+                break;
+        }
```
Should this leverage the `builder.Services.AddMinio` method that is provided by the Minio NuGet package, rather than duplicating the logic?

Harold-Morgan replied:
I was trying to use it, but [it builds the service provider here](https://github.com/minio/minio-dotnet/blob/8daf3d8009180a196798a5b766b2da5bc31835e9/Minio/ServiceCollectionExtensions.cs#L48) and that  interferes with delay validation of connection information [which is expressed in this Conformance Test Failing :)](https://github.com/CommunityToolkit/Aspire/blob/01c80b40a0c847f65fb867de56354126931461ea/tests/CommunityToolkit.Aspire.Testing/ConformanceTests.cs#L356)

aaronpowell replied:
All good then

aaronpowell commented on src/CommunityToolkit.Aspire.Minio.Client/README.md:
````diff
@@ -0,0 +1,119 @@
+In the _Program.cs_ file of your project, call the `AddMinioClient` extension method to register a `MinioClient` for use via the dependency injection container. The method takes a connection name parameter.
+
+```csharp
+builder.AddMinioClient();
````
```suggestion
builder.AddMinioClient("minio");
```

aaronpowell commented on tests/CommunityToolkit.Aspire.Hosting.Minio.Tests/AppHostTests.cs:
```diff
@@ -0,0 +1,43 @@
+        Assert.Equal(HttpStatusCode.OK, createResponse.StatusCode);
+    
+        var getResponse = await httpClient.GetAsync($"/buckets/{bucketName}");
+        Assert.Equal(HttpStatusCode.OK, getResponse.StatusCode);
```
Can we assert on the returned value, not just the status code?

aaronpowell commented on tests/CommunityToolkit.Aspire.Hosting.Minio.Tests/AppHostTests.cs:
```diff
@@ -0,0 +1,43 @@
+    }
+    
+    [Fact]
+    public async Task ApiServiceCreateData()
```
This test has three separate waiting timeouts, do we really want to wait a potential 15 minutes for the timeout if something is wrong? Could we not instead have a cancellation token that is shared between each call, and maybe reduce the duration?

aaronpowell reviewed (changes requested):
Looks pretty good. Got a few changes I've noted.

Also, the MinIO docs have the `IO` part in all capitals (as does the NuGet package), but there's inconsistency in the readme/doc comments - can we have that updated to match the product naming please.

Harold-Morgan commented:
I would definitely be more than happy to be added as a long-term owner of the integration. 

I was planning on adding features and solving issues in the future myself, it is a wonderful opportunity to learn!

@aaronpowell I must note that I haven't yet addressed all the comments in this review (but definitely plan to do so), but you have approved it. Is this ok?

aaronpowell commented:
Sorry, that was a mistake on the approval - notifications took me to only seeing some of the changes and I forgot I had more on the list outstanding so I've rolled it back to waiting approval.

max2020204 commented:
When it can be available from NuGet package?

aaronpowell commented:
> When it can be available from NuGet package?

A preview package should be on NuGet soon (takes a few days to go through approval), but you can always use the Azure DevOps feed as [described on the README](https://github.com/CommunityToolkit/Aspire?tab=readme-ov-file#-installation)

ctolkien commented:
@Harold-Morgan  - thanks for doing this!!

davidfowl commented:
Really high quality work @Harold-Morgan !

Harold-Morgan commented:
> Really high quality work @Harold-Morgan !

![j2hkcly5xc781.png](https://github.com/user-attachments/assets/8756932b-8c0b-413d-b1ba-443b984ce550)



Jokes aside, sincerely thank you and @aaronpowell for help!

I hope and plan to do more
